### ISSUE TITLE
[BUG] Running in a notebook requires installing `ipywidgets`

### ISSUE THREAD

**Issue by gsganden** [ShopRunner/octopod#70]
**Describe the bug**

Functions that create progress bars raise an error in a Jupyter notebook unless `ipywidgets` is installed.

I wouldn't quite call this behavior a bug; it's a limitation of a notebook environment in which `ipywidgets` is not installed, rather than a problem in Octopod per se. We could take the stance that it isn't Octopod's job to handle this issue. That said, I suggest that the benefits of addressing this issue by adding `ipywidgets` to Octopod's requirements outweigh the costs.

- Benefit: Notebook users (who I suspect are in the majority) can install Octopod and run it right off the bat without a hitch.
- Cost: Octopod has another dependency, which increases its footprint and introduces another way for it to break, but it's a relatively lightweight dependency from a reputable source.

**Expected behavior**

Functions that create progress bars work in a Jupyter notebook without additional installation steps.

**To Reproduce**

Run this code in a notebook after installing `octopod` in a clean environment:

```
from octopod.vision.models import ResnetForMultiTaskClassification


model = ResnetForMultiTaskClassification(
    new_task_dict={'fake_task': 2},
    load_pretrained_resnet=True
)
```

**Environment**

Here's a process to create an environment in which this issue occurs.  Create this Dockerfile:

```
FROM python:3.8
LABEL maintainer="data-science@shoprunner.com"
RUN apt-get update \
    && apt-get install -y ffmpeg libsm6 libxext6

RUN pip install jupyterlab==3.0.7 octopod==3.1.6

CMD jupyter lab --allow-root --ip 0.0.0.0 --no-browser
```

Run this command:

```
docker build . -t test
docker run -it -p 8888:8888 test
```

Connect to the notebook server running on port 8888 and run the code above.

**Any additional information?**

Error traceback:

```
Downloading: "https://download.pytorch.org/models/resnet50-19c8e357.pth" to /root/.cache/torch/hub/checkpoints/resnet50-19c8e357.pth
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Input In [1], in <module>
      1 from octopod.vision.models import ResnetForMultiTaskClassification
      4 new_task_dict = {'fake_task':2}
----> 6 model = ResnetForMultiTaskClassification(
      7     new_task_dict=new_task_dict,
      8     load_pretrained_resnet=True
      9 )

File /usr/local/lib/python3.8/site-packages/octopod/vision/models/multi_task_resnet.py:55, in ResnetForMultiTaskClassification.__init__(self, pretrained_task_dict, new_task_dict, load_pretrained_resnet)
     52 def __init__(self, pretrained_task_dict=None, new_task_dict=None, load_pretrained_resnet=False):
     53     super(ResnetForMultiTaskClassification, self).__init__()
---> 55     self.resnet = torch_models.resnet50(pretrained=load_pretrained_resnet)
     56     self.resnet.fc = _Identity()
     58     self.dense_layers = nn.Sequential(
     59         _dense_block(2048*2, 1024, 2e-3),
     60         _dense_block(1024, 512, 2e-3),
     61         _dense_block(512, 256, 2e-3),
     62     )

File /usr/local/lib/python3.8/site-packages/torchvision/models/resnet.py:188, in resnet50(pretrained, **kwargs)
    186 model = ResNet(Bottleneck, [3, 4, 6, 3], **kwargs)
    187 if pretrained:
--> 188     model.load_state_dict(model_zoo.load_url(model_urls['resnet50']))
    189 return model

File /usr/local/lib/python3.8/site-packages/torch/hub.py:586, in load_state_dict_from_url(url, model_dir, map_location, progress, check_hash, file_name)
    584         r = HASH_REGEX.search(filename)  # r is Optional[Match[str]]
    585         hash_prefix = r.group(1) if r else None
--> 586     download_url_to_file(url, cached_file, hash_prefix, progress=progress)
    588 if _is_legacy_zip_format(cached_file):
    589     return _legacy_zip_load(cached_file, model_dir, map_location)

File /usr/local/lib/python3.8/site-packages/torch/hub.py:471, in download_url_to_file(url, dst, hash_prefix, progress)
    469 if hash_prefix is not None:
    470     sha256 = hashlib.sha256()
--> 471 with tqdm(total=file_size, disable=not progress,
    472           unit='B', unit_scale=True, unit_divisor=1024) as pbar:
    473     while True:
    474         buffer = u.read(8192)

File /usr/local/lib/python3.8/site-packages/tqdm/notebook.py:242, in tqdm_notebook.__init__(self, *args, **kwargs)
    240 unit_scale = 1 if self.unit_scale is True else self.unit_scale or 1
    241 total = self.total * unit_scale if self.total else self.total
--> 242 self.container = self.status_printer(self.fp, total, self.desc, self.ncols)
    243 self.container.pbar = proxy(self)
    244 self.displayed = False

File /usr/local/lib/python3.8/site-packages/tqdm/notebook.py:115, in tqdm_notebook.status_printer(_, total, desc, ncols)
    106 # Fallback to text bar if there's no total
    107 # DEPRECATED: replaced with an 'info' style bar
    108 # if not total:
   (...)
    112 
    113 # Prepare IPython progress bar
    114 if IProgress is None:  # #187 #451 #558 #872
--> 115     raise ImportError(
    116         "IProgress not found. Please update jupyter and ipywidgets."
    117         " See [https://ipywidgets.readthedocs.io/en/stable](https://ipywidgets.readthedocs.io/en/stable%3C/span%3E%3Cspan)[ style="color:rgb(175,0,0)">"
    118         "/user_install.html")
    119 if total:
    120     pbar = IProgress(min=0, max=total)

ImportError: IProgress not found. Please update jupyter and ipywidgets. See ](https://ipywidgets.readthedocs.io/en/stable%3C/span%3E%3Cspan)https://ipywidgets.readthedocs.io/en/stable/user_install.html
```

**Comment by gsganden** [comment]
@sugi-chan @parsing-science would you like me to make a PR adding `ipywidgets` to the requirements in `setup.py`?

**Comment by sugi-chan** [comment]
I can try look at it this weekend... I currently dont have a mac tho (i broke my work laptop lol) and forget how annoying docker is on windows... I can look into it...

**Comment by parsing-science** [comment]
Works for me @gsganden if you want to do it or if @sugi-chan does. I can review a PR.

**Comment by gsganden** [comment]
> I can try look at it this weekend... I currently dont have a mac tho (i broke my work laptop lol) and forget how annoying docker is on windows... I can look into it...

Np, no rush. You could probably get the same result by installing `octopod` and some form of Jupyter in a fresh virtual environment, but the Docker route should be more reliable.

**Comment by gsganden** [comment]
> You could probably get the same result by installing octopod and some form of Jupyter in a fresh virtual environment

Actually when I tried this approach just now I didn't get an error 🤷

**Comment by gsganden** [comment]
Closed by #71
